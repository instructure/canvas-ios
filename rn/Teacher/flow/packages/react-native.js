//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
