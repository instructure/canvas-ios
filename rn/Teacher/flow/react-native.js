//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
