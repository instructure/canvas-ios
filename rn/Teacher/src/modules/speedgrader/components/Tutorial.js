//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { PureComponent } from 'react'
import {
  AsyncStorage,
  View,
  Image,
  StyleSheet,
  Animated,
  NativeModules,
} from 'react-native'
import { LinkButton } from '../../../common/buttons'
import { Text } from '../../../common/text'
import i18n from 'format-message'

const { NativeAccessibility } = NativeModules

type TutorialType = {
  id: string,
  text: string,
  image: any,
}

type Props = {
  tutorials: Array<TutorialType>,
}

type State = {
  hasLoaded: boolean,
  hasSeen: { [string]: boolean },
  currentTutorial: ?TutorialType,
}

const SPEED_GRADER_TUTORIAL_KEY = 'speed-grader-tutorial'

export default class Tutorial extends PureComponent<Props, State> {
  state: State = {
    hasLoaded: false,
    hasSeen: {},
    currentTutorial: null,
  }
  opacity: Animated.Value = new Animated.Value(0)

  componentDidMount () {
    this.setup()
  }

  setup = async () => {
    let data = await AsyncStorage.getItem(SPEED_GRADER_TUTORIAL_KEY)
    let hasSeen = JSON.parse(data || '{}')

    this.setState({
      hasLoaded: true,
      hasSeen,
      currentTutorial: this.props.tutorials.find(({ id }) => !hasSeen[id]),
    })

    Animated.timing(
      this.opacity,
      {
        toValue: 1,
      }
    ).start(this.focusCurrentTutorial)
  }

  onPress = async () => {
    let currentTutorial = this.state.currentTutorial
    if (!currentTutorial) return null

    let hasSeen = {
      ...this.state.hasSeen,
      [currentTutorial.id]: true,
    }
    AsyncStorage.setItem(SPEED_GRADER_TUTORIAL_KEY, JSON.stringify(hasSeen))
    let nextTutorial = this.props.tutorials.find(({ id }) => !hasSeen[id])

    await new Promise(resolve => {
      if (nextTutorial) return resolve()
      Animated.timing(
        this.opacity,
        {
          toValue: 0,
        }
      ).start(() => {
        this.focusCurrentTutorial()
        resolve()
      })
    })

    this.setState({
      hasSeen,
      currentTutorial: nextTutorial,
    })
  }

  focusCurrentTutorial = () => {
    if (this.state.currentTutorial) {
      const current = this.state.currentTutorial
      const elementID = `${current.id}-title`
      setTimeout(() => {
        NativeAccessibility.focusElement(elementID)
      }, 500)
    }
  }

  render () {
    let currentTutorial = this.state.currentTutorial
    if (!this.state.hasLoaded) return null
    if (!currentTutorial) return null

    return (
      <Animated.View
        accessibilityViewIsModal={true}
        style={[
          styles.tutorial,
          {
            opacity: this.opacity,
          },
        ]}
      >
        <View style={styles.box}>
          <Text style={styles.text} testID={`${currentTutorial.id}-title`}>{currentTutorial.text}</Text>
          <Image style={styles.image} source={currentTutorial.image} />
          <LinkButton
            textStyle={styles.buttonText}
            onPress={this.onPress}
            hitSlop={{ top: 10, left: 10, right: 10, bottom: 10 }}
            underlayColor='#fff'
            testID={`tutorial.button-${currentTutorial.id}`}
          >
            {i18n('OK')}
          </LinkButton>
        </View>
      </Animated.View>
    )
  }
}

const styles = StyleSheet.create({
  tutorial: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
  },
  box: {
    backgroundColor: 'white',
    borderRadius: 8,
    padding: 16,
    marginHorizontal: 40,
    shadowColor: 'black',
    shadowRadius: 10,
    shadowOpacity: 0.5,
    shadowOffset: { width: 0, height: 10 },
  },
  text: {
    fontWeight: '600',
    fontSize: 18,
    marginBottom: 16,
  },
  buttonText: {
    fontSize: 16,
    fontWeight: '600',
    textAlign: 'center',
  },
  image: {
    marginBottom: 24,
  },
})
