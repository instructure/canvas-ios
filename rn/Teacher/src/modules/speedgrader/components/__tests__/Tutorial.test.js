//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { Animated, NativeModules } from 'react-native'
import renderer from 'react-test-renderer'
import Tutorial from '../Tutorial'
import Images from '../../../../images'
import explore from '../../../../../test/helpers/explore'

const { NativeAccessibility } = NativeModules

jest
  .mock('@react-native-community/async-storage', () => ({
    getItem: jest.fn(() => Promise.resolve()),
    setItem: jest.fn(() => Promise.resolve()),
  }))
  .mock('react-native/Libraries/Animated/src/Animated', () => ({
    Value: jest.fn(),
    timing: () => ({ start: jest.fn() }),
    View: 'Animated.View',
  }))
  .mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity')

let defaultProps = {
  tutorials: [{
    id: '1',
    text: 'Some text',
    image: Images.speedGrader.swipe,
  }],
}

describe('Tutorial', () => {
  beforeEach(() => jest.clearAllMocks())
  it('will render null before data from async storage has been retrieved', () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    ).toJSON()

    expect(tree).toBeNull()
  })

  it('will render the tutorial after data from async storage has been retrieved', () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )
    tree.getInstance().setState({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: defaultProps.tutorials[0],
    })

    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('will figure out what tutorial to show in setup', async () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )

    await tree.getInstance().setup()

    expect(tree.getInstance().state).toEqual({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: defaultProps.tutorials[0],
    })
  })

  it('will update hasSeen when onPress is called', async () => {
    let oldTiming = Animated.timing
    let start = jest.fn()
    Animated.timing = () => ({ start })

    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )
    tree.getInstance().setState({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: defaultProps.tutorials[0],
    })

    let button = explore(tree.toJSON()).selectByID('tutorial.button-1') || {}
    let promise = button.props.onPress()

    start.mock.calls[0][0]()
    await promise

    expect(tree.getInstance().state).toEqual({
      hasLoaded: true,
      hasSeen: { '1': true },
      currentTutorial: undefined,
    })

    Animated.timing = oldTiming
  })

  it('focus current tutorial function should work', () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )
    const currentTutorial = defaultProps.tutorials[0]
    tree.getInstance().setState({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial,
    })

    jest.useFakeTimers()
    tree.getInstance().focusCurrentTutorial()
    jest.runAllTimers()
    expect(NativeAccessibility.focusElement).toHaveBeenCalledWith(`${currentTutorial.id}-title`)
  })

  it('focus current tutorial function should not do anything when there is no current tutorial', () => {
    let tree = renderer.create(
      <Tutorial {...defaultProps} />
    )
    tree.getInstance().setState({
      hasLoaded: true,
      hasSeen: {},
      currentTutorial: null,
    })

    tree.getInstance().focusCurrentTutorial()
    expect(NativeAccessibility.focusElement).not.toHaveBeenCalled()
  })
})
