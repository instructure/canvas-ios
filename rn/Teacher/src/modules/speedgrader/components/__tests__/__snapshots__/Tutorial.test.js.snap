// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tutorial will render the tutorial after data from async storage has been retrieved 1`] = `
<View
  accessibilityViewIsModal={true}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.5)",
      "bottom": 0,
      "display": "flex",
      "justifyContent": "center",
      "left": 0,
      "opacity": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "borderRadius": 8,
        "marginHorizontal": 40,
        "padding": 16,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 10,
          "width": 0,
        },
        "shadowOpacity": 0.5,
        "shadowRadius": 10,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "fontFamily": ".SFUIDisplay",
          },
          Object {
            "color": "#2D3B45",
            "fontSize": 16,
          },
          Object {
            "fontSize": 18,
            "fontWeight": "600",
            "marginBottom": 16,
          },
          Object {
            "fontFamily": ".SFUIDisplay-semibold",
          },
        ]
      }
      testID="1-title"
    >
      Some text
    </Text>
    <Image
      source={
        Object {
          "testUri": "../../../src/images/speedgrader/Swipe.png",
        }
      }
      style={
        Object {
          "marginBottom": 24,
        }
      }
    />
    <TouchableOpacity
      accessibilityTraits="button"
      hitSlop={
        Object {
          "bottom": 10,
          "left": 10,
          "right": 10,
          "top": 10,
        }
      }
      onPress={[Function]}
      testID="tutorial.button-1"
      underlayColor="#fff"
    >
      <View>
        <Text
          style={
            Array [
              Object {
                "fontFamily": ".SFUIDisplay",
              },
              Object {
                "color": "#2D3B45",
                "fontSize": 16,
              },
              Array [
                Object {
                  "color": "#374A59",
                  "fontSize": 16,
                  "fontWeight": "500",
                },
                Object {
                  "fontSize": 16,
                  "fontWeight": "600",
                  "textAlign": "center",
                },
              ],
              Object {
                "fontFamily": ".SFUIDisplay-semibold",
              },
            ]
          }
        >
          OK
        </Text>
      </View>
    </TouchableOpacity>
  </View>
</View>
`;
